### PR TITLE
[NFCI][SYCL] Rework of DeclBodyCreator/Array Handling in prep of non-Decomposition

### PR DESCRIPTION
**Pull request by erichkeane**
The intent of this patch is simply to clean up how arrays work in the
SYCL visitors.  The patch unifies the handling of array elements and 
normal fields, such that they should be more maintainable.

This required a handful of changes to make integration headers work
a little better.

Work on this DID require quite a bit of a re-imagining of how the
DeclBodyCreator works. There are a number of small refactors to fix
things I found while working that I believe make it easier to maintain
this in the future.

However, the biggest change here is with the init-list-expr creation.
This patch creates them in advance, rather than waiting until we are
done with a struct/array/etc.  This allows us to generate them as we go.

The side effect of this is that the InitExprs array now ONLY contains
collection InitListExprs, since it no longer has to assemble them on the
other side.  However, now adding a field ends up being really easy,
since appending to the init-list is easier.

MemberExprBases also becomes easier to maintain, since they now only
get added when changing to/from a field or collection.

### CODE REVIEW

**Comment by erichkeane** [comment]
Note, this is a draft PR, just enough so that you two can see what I chose to do, and question my design decisions :)

**Comment by erichkeane** [comment]
Also just discovered that arrays of structs isn't working correctly, since it only has the FieldDecl, instead of its actual type for enter-struct.   So there IS some work in the visitor left :)

**Comment by erichkeane** [comment]
There, got arrays of structs and special-type base class initialization working right.  Last 4 failures are:
Failed Tests (4):
  Clang :: CodeGenSYCL/kernel-param-acc-array.cpp
  Clang :: CodeGenSYCL/kernel-param-member-acc-array.cpp
  Clang :: CodeGenSYCL/stream.cpp
  Clang :: SemaSYCL/array-kernel-param.cpp

I'm sure I have a bunch of array problems with the MemberExprBases, since I haven't touched those yet.  I know I also have a problem with the 'stream' type since I haven't touched it very much, but this is all I'm going to do for the evening.

**Comment by erichkeane** [comment]
@Fznamznon and @elizabethandrews , can you pay particular attention to 
77b12c7  (https://github.com/intel/llvm/pull/2370/commits/77b12c721f2b9ea57ebab1151d83b9aae76d9231 Fix Array Kernel Param Test)?

I'm reasonably sure that the test was wrong (and that we had the wrong behavior!), but I want to make sure since this is a change in behavior.

**Comment by erichkeane** [comment]
> 
> 
> I have not had the chance to look at your patch yet. I will look at it today

At the moment, it is an "FYI".  I'm pretty close to getting it passing all the the CFE tests, but likely have a few hours of cleanup to do as well (to deal with/think about all the TODOs I have around).  This is mostly to make sure this isn't a surprise to you if you'd like to take a look.

The three of us can have a quick call at one point if the two of you want an explaination of what I am doing.  @premanandrao  is welcome too :)

**Comment by erichkeane** [comment]
Alright, all the LIT tests pass for me, we'll see what the buildbots have to say :)  Two of the tests for accessors in arrays seem to have changed, but I believe they are correct (and further, I think in at least 1 of them, it was wrong before, as it was accessing index2 of a 2 item array!).  I hopefully improved the test.

**Comment by erichkeane** [comment]
Hm... looks like I picked up 3 failures :/  @elizabethandrews or @Fznamznon  Any chance you can help me take a look?

**Comment by Fznamznon** [comment]
> Hm... looks like I picked up 3 failures :/ @elizabethandrews or @Fznamznon Any chance you can help me take a look?

Hmm, all failing tests use stream. I'll take a look, but I will be offline in a couple of hours and do not guarantee that I will give something working before I go offline. The first thing I would do is compare llvm-ir for these tests before and after your changes. If your change is NFC change against stream handling it should be the same.

**Comment by Fznamznon** [comment]
> > Hm... looks like I picked up 3 failures :/ @elizabethandrews or @Fznamznon Any chance you can help me take a look?
> 
> Hmm, all failing tests use stream. I'll take a look, but I will be offline in a couple of hours and do not guarantee that I will give something working before I go offline. The first thing I would do is compare llvm-ir for these tests before and after your changes. If your change is NFC change against stream handling it should be the same.

I know what happened. I see that __init call for stream class is generated before __init calls for accessors which are wrapped by stream. I think it breaks stream because __init of stream actually uses one of internal stream accessors, see https://github.com/intel/llvm/blob/e5036625b05937d2961dea1d9b729a50455c5fd5/sycl/include/CL/sycl/stream.hpp#L819 (GlobalOffset is accessor).
It would be great to re-implement stream class one time, so we don't have to handle it specially in compiler (see https://github.com/intel/llvm/pull/2268).

**Comment by erichkeane** [comment]
> 
> 
> > > Hm... looks like I picked up 3 failures :/ @elizabethandrews or @Fznamznon Any chance you can help me take a look?
> > 
> > 
> > Hmm, all failing tests use stream. I'll take a look, but I will be offline in a couple of hours and do not guarantee that I will give something working before I go offline. The first thing I would do is compare llvm-ir for these tests before and after your changes. If your change is NFC change against stream handling it should be the same.
> 
> I know what happened. I see that __init call for stream class is generated before __init calls for accessors which are wrapped by stream. I think it breaks stream because __init of stream actually uses one of internal stream accessors, see
> 
> https://github.com/intel/llvm/blob/e5036625b05937d2961dea1d9b729a50455c5fd5/sycl/include/CL/sycl/stream.hpp#L819
> (GlobalOffset is accessor).
> It would be great to re-implement stream class one time, so we don't have to handle it specially in compiler (see #2268).

Ah, you're awesome!  Thanks! 

I didn't think that would cause a problem, one thing that DID change was that the 'stream' body handling was moved to be before going down its members.  This was to make it more consistent with the other special types/collections, but I didn't even think that this would matter. 

This is at least easy to solve, and it is a relief that the failures are all stream releated :)  

I'll get a new patch up to fix that later today as I look into some of the TODOs that I listed (of which, there are 5, 3 of which I think I should deal with, the other two are perhaps for you/Elizabeth).

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1067]: No, it does not.  addFieldInit handles cases where it is in an init.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1932]: No, the 'addBaseInit' handles that properly it seems.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1781]: I think I split the code that made this true into a separate location, so it is mostly true.  Unless someone has a suggestion, I think this TODO is going away.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:1783]: Copy ends up being more generic, the each have specific issues based on what is coming in.  We could potentially optimize the init at one point by making this one of the other two, but it would have to be on a per-type basis.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2058]: This is actually  necessary, it sets up casts correctly.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2269]: This IS wrong thanks to my refactor... Still working through some issues on it, but this is my last issue as far as I know.

**Comment by erichkeane** [comment]
Great, now I seem to have broken arrays :/  Probably my int-header issue, at least the array-kernel-param-run.cpp actually PRINTS stuff of value!

**Comment by erichkeane** [comment]
@bader  IS there a problem with the buildbots? I've got two builds in a row (both of which shouldn't have had issues as far as I can tell) where the build command hits a timeout.

command timed out: 1200 seconds without output running [b'python3', b'llvm_ci/intel/worker/tools/build.py', b'-n', b'4702', b'-b', b'pull/2370/head', b'-r', b'2370', b'-t', b'check-sycl', b'-p', b'sycl', b'-s', b'lit', b'-P', b'intel/llvm', b'-m', b'sycl-ubu-x64-pr', b'-e', b'a6ee42509b41bd7e149eba8758d0f38a3097699e', b'-U', b'http://ci.llvm.intel.com:8010/#/builders/2/builds/4702'], attempting to kill

It seems that this is way too short to run our tests...

**Comment by Fznamznon** [comment]
> @bader IS there a problem with the buildbots? I've got two builds in a row (both of which shouldn't have had issues as far as I can tell) where the build command hits a timeout.
> 
> command timed out: 1200 seconds without output running [b'python3', b'llvm_ci/intel/worker/tools/build.py', b'-n', b'4702', b'-b', b'pull/2370/head', b'-r', b'2370', b'-t', b'check-sycl', b'-p', b'sycl', b'-s', b'lit', b'-P', b'intel/llvm', b'-m', b'sycl-ubu-x64-pr', b'-e', b'a6ee42509b41bd7e149eba8758d0f38a3097699e', b'-U', b'http://ci.llvm.intel.com:8010/#/builders/2/builds/4702'], attempting to kill
> 
> It seems that this is way too short to run our tests...

AFAIK we have a team working on infrastructure. Tagging @DoyleLi , @tfzhu .

**Comment by erichkeane** [comment]
I updated the commit message, I believe this is ready for review so I'm marking it as such.  I presume we'll have a couple of cycles of review, so hopefully the buildbots will be better by then.

**Comment by erichkeane** [comment]
Ah, I see lost in the linux timeout, the Windows build seems to have a problem with the 'stream.cpp' still.  I don't really see any IR/int-header differences there that look significant, but does @elizabethandrews or @Fznamznon perhaps have any thoughts/help they could give?

**Comment by erichkeane** [comment]
Huh... actually, capturing 2 streams results in an assert, though that doesn't show in the test.  I'll have to look closer.

**Comment by erichkeane** [comment]
I found the (admittedly foolish :/) issue I made when splitting streams into a separate set of enter/leave.  I have it fixed locally, but with as many problems as we've had with streams I'm trying to write a fairly comprehensive test here.

**Comment by Fznamznon** [comment]
> I have it fixed locally, but with as many problems as we've had with streams I'm trying to write a fairly comprehensive test here.

It can fix https://github.com/intel/llvm/issues/1553

**Comment by erichkeane** [comment]
> 
> 
> > I have it fixed locally, but with as many problems as we've had with streams I'm trying to write a fairly comprehensive test here.
> 
> It can fix #1553

Well, it at least fixes the AST part of that problem :)  There is probably a number of other tests that could be written, but the coverage for how streams are generated into AST boht in/out of arrays and structs have been done.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:2197]: Array is the only case where handler function receives type and field of other type, right?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:899]: I feel we have some problems with code style in lots of places

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:1663]: It is not obvious what uint64_t part of pair holds, could you please add a comment?

**Comment by erichkeane** [comment]
@Fznamznon : This is ready for review!  The test mentioned above is in 
7ed04d7

**Review by Fznamznon — COMMENTED** [review]
This overall looks better than it was before, but still requires some time to understand what is going on.
In general I'm fine with changes. There are a lot of code-style problems, but I'm not sure if I should go and create a comment for each one. In addition I'm not sure that stream testing should go together with this patch.
Please note that I ran comprehensive testing and I prefer waiting for results of it before merge, because it is significant refactoring and I want to make sure that we don't revert it after commit.

**Comment by Fznamznon** [comment]
BTW, since it is a preparation for non-decomposition, I'd like to hear a feedback from @elizabethandrews as well.

File: clang/test/SemaSYCL/Inputs/sycl.hpp
Comment (erichkeane) [reply clang/test/SemaSYCL/Inputs/sycl.hpp:221]: I can add one... it makes the test that much longer.

I wrote this test as a part of this patch, since it actually manages to test all of the funny 'edge cases' I could think of to handle structs/arrays. It helped me find a handful of bugs in my implementation.

I can remove it from the patch if you'd like, I just thought it was a useful addition.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2197]: Yes, exactly.  It is the only case where the FieldDecl doesn't represent the thing that we're visiting.  I would like some day to come up with a better way to differentiate, but its not obvious to me how to do this without some sort of state being set in the visitor.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:899]: This is after clang-format, can you clarify?

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:899]: I meant problems with names, i.e. function names should start with lower case letter, variable names start with upper case letter and so on. clang-format doesn't care about it.

File: clang/test/SemaSYCL/Inputs/sycl.hpp
Comment (Fznamznon) [reply clang/test/SemaSYCL/Inputs/sycl.hpp:221]: But without this accessor we will have to figure out problems like we had with __init calls sequence again.

**Comment by erichkeane** [comment]
> 
> 
> This overall looks better than it was before, but still requires some time to understand what is going on.
> In general I'm fine with changes. There are a lot of code-style problems, but I'm not sure if I should go and create a comment for each one. In addition I'm not sure that stream testing should go together with this patch.
> Please note that I ran comprehensive testing and I prefer waiting for results of it before merge, because it is significant refactoring and I want to make sure that we don't revert it after commit.


Unfortunately, this ended up being one of those patches that required changing a lot of things in order to set everything right.  I'd offered above to go through some of the changes with you, @elizabethandrews  and @premanandrao if you'd like!

However, mostly, the DeclBodyCreator was switched to creating the InitListExpr upon ENTERING a collection, rather than trying to set it up afterwards.  Array init-list-exprs didn't like being updated as we went, and was difficult to keep track of (hence the massive amount of hackery around the MemberBaseExprs).  

Everything else falls out of that :)  Part of THAT was unifying the order of operations in a few places (making sure handleSyclStream happened before its elements).

File: clang/test/SemaSYCL/Inputs/sycl.hpp
Comment (erichkeane) [reply clang/test/SemaSYCL/Inputs/sycl.hpp:221]: Thats a good point, I can add an accessor here.

**Comment by Fznamznon** [comment]
> I'd offered above to go through some of the changes with you, @elizabethandrews and @premanandrao if you'd like!

I don't mind, that would be helpful.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:899]: Ah, I see.  I'll make a run through this patch myself, then have you correct anything I missed.

**Comment by erichkeane** [comment]
Ok, I should now have the streams test with the additional accessor init calls, plus fixed all the wrong function names that I added.  I didn't see any variable issues/etc, but please let me know what else you see!

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:840]: I guess "handlers" also should start with upper case letter...

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:869]:
```suggestion
        visitRecord(Owner, Base, BaseTy->getAsCXXRecordDecl(), BaseTy,
```

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:1832]:
```suggestion
  MemberExpr *buildMemberExpr(Expr *Base, ValueDecl *Member) {
```

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:883]: Yeah, there are a lot of them outside of your changes :(
```suggestion
  void visitStreamRecord(const CXXRecordDecl *Owner, ParentTy &Parent,
```

**Review by Fznamznon — COMMENTED** [review]
Feel free to leave nits that are outside of your changes as is.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:1916]: Can you explain what is happening here? I understand you're updating ParentILE with the 'new' empty initializer returned in L1909 but it doesn't look like you're saving this change? ParentILE isn't used after this. Do you need to update CollectionInitListExpr?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp]: Note that I couldn't just capitalize Handlers, since that was taken by the pack.  I changed the pack to HanderTys, and the variables to Handlers.  I did this throughout the visitor type.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:1916]: Exactly what you said.  This creates a new init-list, and adds it to the current collection init-expr.  ParentILE is just the last element of the CollectionInitExprs, so it is already there, there isn't anything left to update.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Yeah, that is why I didn't make this comment in form of dummy suggestion.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1916]: Oh right. It's updated in CollectionInitExprs as well.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:2202]: For pointers, the address space qualifier can be different for FD and the type we generated in DeclCreator. I assume that doesn't matter here?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2202]: Right, it shouldn't matter.  The isArrayElement check happens on the type that we get from the visitor.

**Review by Fznamznon — APPROVED** [review]

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:2253]: Can you add a comment explaining why this specialization was needed?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:2356]: So CurOffset holds the offset of each record we're currently visiting except when inside an array. When it is inside array, it holds offset of each element. When you leave array you revert it back? Did I understand that right?

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:1677]: That works.  I originally thought we would care about being in an array (array of pointers), but it doesn't seem we care otherwise.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:1817]: There WAS a TODO and I think it was right, it isn't STRICTLY correct.  That said, it seems to work?  I tried a while to come up with a case where it wouldn't in order to justify keeping the whole stack of entities (rather than just the arrays), but couldn't come up with one.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2356]: Correct!

File: clang/lib/Sema/SemaSYCL.cpp
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:2258]: Done!

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1817]: I wonder if we'll need to once we stop decomposing classes. If kernel has a field with a base class I assume we can't initialize with VarEntity then right?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:1817]: Thats an interesting question... I don't see why we would need to in the non-decomp case, since those seem like a subset of this case.  But I am sure we'll see when we start the work on non-decomp.

IF thats the case, I suspect making ArrayInfos be simply CollectionInfos would be on the docket.

**Review by Fznamznon — APPROVED** [review]

File: clang/lib/Sema/SemaSYCL.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1817]: >>> since those seem like a subset of this case
Maybe? Right now addBaseInit is called only when special types are used as a base class. I can't recall if we have tests which check if they can be used as base classes of fields of kernel as opposed to base class of functor itself. I understand the latter working but I'm not sure why the former works?

```
class A { }
class B : public A { }
class KernelFunctor {
       B obj;
}
```
VarEntity here is Entity for KernelFunctor object right? I'm not sure how using VarEntity when creating InitializedEntity for A works? If our special bases now works for similar code though, yay! I assume it would work for non-decomp case as well even if I don't understand why it works :)

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:1817]: Right, it is.  I think these work, and I think they'll work for the non-decomp case (though, if I find a good example of it being necessary while implementing non-decomp, I'll obviously do it) for the same reason they work for the special types:  It seems that this entity isn't particularly important except in array cases.

I've got this in my notes as something to pay attention to in the near future as a part of the non-decomp patch/patches, so hopefully we'll know then.

**Comment by erichkeane** [comment]
@pvchupin  @bader  The jenkins/summary bot seems stuck, it keeps timing out at ~20 hours on a number of patches, so I think it is broken and this patch can go in anyway.  Thoughts?

File: clang/test/SemaSYCL/streams.cpp
Comment (bader) [inline clang/test/SemaSYCL/streams.cpp:9]: I think we can use `single_task` from included header instead.

File: clang/test/SemaSYCL/streams.cpp
Comment (bader) [inline clang/test/SemaSYCL/streams.cpp:3]: It looks like other tests use following pattern:
```c++
#include "Inputs/header.hpp"
```
instead of 
```c++
// RUN: %clang_cc1 -I %S/Inputs ...
#include "header.hpp"
```
I suggest we refactor our SYCL LIT tests to follow the patter w/o additional compiler options.
I can create a separate PR, if there are no objections.

File: clang/test/SemaSYCL/streams.cpp
Comment (bader) [inline clang/test/SemaSYCL/streams.cpp:1273]: 1200 lines of checks... WOW.

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
> @pvchupin @bader The jenkins/summary bot seems stuck, it keeps timing out at ~20 hours on a number of patches, so I think it is broken and this patch can go in anyway. Thoughts?

I'm okay, considering that this is supposed to be non-functional change.

File: clang/test/SemaSYCL/streams.cpp
Comment (erichkeane) [reply clang/test/SemaSYCL/streams.cpp:3]: I don't see a problem with it.  Getting them all the same would be nice.  In this case, I ended up copy/pasting the cc1 invocation/include from another file, so there's definitely more like it that have the same problem.

**Comment by erichkeane** [comment]
FYI, looks like the tests finally passed!

**Pull request merged**
